tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -759,29 +760,50 @@ public String toString() {
       sb.append(String.format("[Index: %s] ", indexInfo.getName()));
     }
 
-    if (getFields().size() != 0) {
+    if (!getFields().isEmpty()) {
       sb.append(", Columns: ");
-      sb.append(Joiner.on(", ").skipNulls().join(getFields()));
+      Joiner.on(", ").skipNulls().appendTo(sb, getFields());
     }
 
-    if (getFilters().size() != 0) {
-      sb.append(", Filter: ");
-      sb.append(Joiner.on(", ").skipNulls().join(getFilters()));
+    if (!getDowngradeFilters().isEmpty()) {
+      // should be called after all parameters are set
+      if (pushdownFilters == null) {
+        pushdownFilters = new ArrayList<>(getDowngradeFilters());
+        pushdownFilters.removeAll(new HashSet<>(getFilters()));
```

Comment:
Please confirm that all expressions has properly implemented hashcode and equals.

Reply:
All equals are properly implemented. Although hashcode for `Not(child)` and `IsNull(child)` are the same as `child`, they wouldn't be a big deal.